### ISSUE TITLE
Adding new songs replaces existing library with Unknown (Orphaned)

### ISSUE THREAD

**Issue by codemaestro** [ampache/ampache#3534]
## Description

I use version Ampache 5.6.1 provided by Softaculous. I haven't had the time to manually download the `develop` branch to test for this fix. I just assume that since the issue occurred before and including 5.6.0 and persists to 5.6.1 that my use case probably hasn't been considered.

### Describe the bug
For my jam band, I share `m4a` recordings through a privately-hosted Ampache instance. The structure is as follows:

- Defined: `path/to/library/folder`
- Structure: `Folder of songs - Names of players, separated, by commas`
- Song files: `Song title.m4a`

When I first add the catalog and run the `Add Songs` operation, the albums, artists, and songs are catalogued perfectly.

When I add a new folder of songs and run any of `Add`, `Update`, `Verify`, etc. operations, the new album, artist, and songs are catalogued perfectly, but the existing catalog updates all the artist names to `Unknown (Orphaned)`.

### To reproduce

1. On the server, in the music source folder, create a folder for `album title - artist` and upload files.
2. Under `Admin`, click **Add Catalog**.
3. Add the information for a local catalog, including URL and metadata patterns for files and folders.
4. Click **Add Catalog**.
5. After the process completes, click **Home** and go to **Albums**. The catalog looks and works correctly.
6. On the server, add a new `album title - artist` folder and songs, following the same pattern.
7. Under `Admin`, click **Show Catalogs**.
8. Beside the existing catalog, choose any of the options, like **Add**, **Update**, etc.
5. After the process completes, click **Home** and go to **Albums**. The catalog replaces the artist for existing songs to `Unknown (Orphaned)`, but the new album looks and works correctly.

### Expected behavior

The existing songs shouldn't change. Only the new songs should be added in the same manner.

### Screenshots

<img width="1144" alt="image" src="https://github.com/ampache/ampache/assets/1635612/eba41970-6ee3-445a-a4b8-8b9a2f88f406">

* Ampache version: 5.6.1
* Web server + version: Hosted system

**Comment by lachlan-00** [comment]
So when you add new files the album artist is ok?
Then a clean, add verify gives you the orphans?

If you're able to edit the files we can add more debug to the tag process which would make it easier to process.

Another question is it only the album artist tag that is is getting updated?

**Comment by codemaestro** [comment]
You understand my experience exactly. When new files are added, they have
the correct album and artist tags. But the existing files are retagged as
orphans. And it’s only the artist tag.

I can edit the files to add whatever debugging info you need.

**Comment by lachlan-00** [comment]
would you by chance not have an albumartist tag in your files?

**Comment by codemaestro** [comment]
When I load the catalog, I use the following pattern:

File name: %T %t

Folder name: %A - %a

The Add routine follows these rules, but the Update routine doesn’t seem
to. Also, there doesn’t seem to be a way to view the tags used to import a
catalog.

**Comment by codemaestro** [comment]
Hi @lachlan-00 - what files would you like me to update to gather the diagnostic details for review?

**Comment by Nizhile** [comment]
I get the same kind of issue with release 5.6.2 and it would probably be the same with the development branch. I use admin account in french. As a consequence, the `Unknown (Orphaned)`  is stored in the artist table as `Inconnu (orphelin)`.
My files are not cleanly stored nor tagged (missing Albumartist tag and no artist name in directory tree).
Anyway, even if Madonna is the single artist of frozen directory album, the Album Artist will stay `Inconnu (orphelin)`.
A look at https://github.com/ampache/ampache/blob/a23ae23aa0383c177f88a1ffb261cdc806a0ed7d/src/Repository/Model/Album.php#L1241 shows that whatever is to be done here, it will probably not work.
As far as I understand, my database content for unknown artist is localized (french) but not the SQL request.

![image](https://github.com/ampache/ampache/assets/11645217/8988e76f-3914-4d64-a380-ef0c947d2fa7)

If an album/directory contains multiple artists, the name of the album artist is not updated (even if I update it manually). This is the case of the New Wave compilation.

**Comment by mezco2** [comment]
Hello!
I am having the exact same problem. Is there a way to manually update the artists' tags?
Thanks.

**Comment by lachlan-00** [comment]
> Unknown (Orphaned)

That's a bug cause it's supposed to be checking for what you've got in your language!

I've set that with d44b32ea7 which will help with other languages that aren't EN-US compatible

**Comment by lachlan-00** [comment]
you can print what you get from tags so we can see whats coming in editing `src/Repository/Model/Catalog.php`

uncomment the debug and update an album from tags to get that printed out.
![image](https://github.com/ampache/ampache/assets/1305249/a9db7dd0-8306-43d7-86ff-e17618e5b64d)

**Comment by Nizhile** [comment]
I’m not sure about the fix for the SQL request. Is the localize function always using the server preferred configuration or the user using the administration interface ? And also, what happens when using cli scripts on the server?
Maybe the database should be using non translated objects (as for date) and only display would be translated?

**Comment by mezco2** [comment]
Hi again,
Firstly, I would like to thank you for your efforts in solving this issue.
Secondly, I should have been clear that my issue is not related to language. My installation is set to US English.
I implemented the change to the catalog.php file noted in d44b32e above, but it made no difference. 
It should be noted that this issue only occurs on the Albums page and at the top of the respective songs page. If I click on the album name, the Song Artist name is correctly listed in the Songs table, but it is not correct on the header at the top of the page.
Pardon my ignorance, but I am not a programmer by any stretch of the imagination. That said, I will do whatever you need me to do to get this problem solved. I just need clear instructions.
To wit: The instructions above state "uncomment the debug and update an album from tags to get that printed out". I apologize, but I do not know how to do that. I assume that removing the "//" from the beginning of the line will uncomment the line? How do I get the print to work from there? I do know how to edit a php file.
I've attached the latest log file. I hope that helps.
Once again, thank you and please don't hesitate to contact me for further info/experimentation.
[ampache.20230824.log](https://github.com/ampache/ampache/files/12433796/ampache.20230824.log)

**Comment by codemaestro** [comment]
As I originally reported this issue, I feel I should weigh in. I am a programmer and while I managed to update the Catalog.php file as instructed, I have no idea where the log files are supposed to be written. I updated the configuration file to move the log from /var/log/ampache (which I can't access on my shared host) to a local file that I can access, but no logs get written.

@mezco2 you got further along than I did by finding the log file!

**Comment by mezco2** [comment]
@codemaestro  -- I managed to get Ampache to create the log file by modifying the ampache.cfg.php file located in the config directory. I uncommented line 826 (debug = "true"), line 838 (debug_level = 5), and set the log path on line 847 to the "logs" directory I created. The full path must be noted in the log path. I found those instructions [here](https://github.com/ampache/ampache/wiki/Troubleshooting#Enable-Logging).
As an FYI, I also installed Ampache through Softaculous and I am also running version 5.6.2 of Ampache.
I hope that helps.

**Comment by Nizhile** [comment]
The localization issue only makes the initial set up incorrect (when creating songs). But I have the same issue. If I update manually album artist, it is reset to Unknown (Orphan) at next catalog update. The issue is linked to ALBUMARTIST tag missing (while album tag is correctly set)
The problem will not appear when tracks have ALBUMARTIST tag. This is why compilation or albums with collaboration does not have the issue.
Mp3tag software will not set ALBUMARTIST is all tracks in the album have the same ARTIST tag (which is expected)

**Comment by Nizhile** [comment]
Actually, the update catalog creates a new album id with an album artist "Unknown Orphan". The original Madonna ray of light with correct album artist is 217 while the new one is 218. Data are migrated from the old one to the newer album without the album artist.
The reason is that the media check changes mode from abr to vbr `media-diff {"album":"OLD: 217 --> 218","mode":"OLD:abr --> vbr"}`. I use some ogg vorbis files without any specific things here and there obviously the same between the create catalog and update catalog operations.
@mezco2 @codemaestro Can you check your debug log files for any `duplicate... type:album` lines?

```
2023-08-25T14:16:27+01:00 [eric] (local.catalog) -> Verify starting on Orphan
2023-08-25T14:16:27+01:00 [eric] (local.catalog) -> catalog 6 starting verify album on chunk 0
2023-08-25T14:16:27+01:00 [eric] (Ampache\Repository\Model\Catalog) -> Reading tags from /home/eric/Testing/orphan/ray_of_light/drowned_world__substitute_for_love.ogg
2023-08-25T14:16:27+01:00 [eric] (Ampache\Module\Util\VaInfo) -> _get_plugin_tags: musicbrainz is not a valid metadata_order plugin
2023-08-25T14:16:27+01:00 [eric] (Ampache\Module\Util\VaInfo) -> _get_plugin_tags: theaudiodb is not a valid metadata_order plugin
2023-08-25T14:16:27+01:00 [eric] (Ampache\Repository\Model\Album) -> check album: created {218}
2023-08-25T14:16:27+01:00 [eric] (Ampache\Repository\Model\Catalog) -> update_map album: {218}
2023-08-25T14:16:27+01:00 [eric] (Ampache\Repository\Model\Album) -> add_album_map album_id {218} song_artist {15}
2023-08-25T14:16:27+01:00 [eric] (Ampache\Repository\Model\Artist) -> add_artist_map artist_id {8} album {218}
2023-08-25T14:16:27+01:00 [eric] (Ampache\Repository\Model\Album) -> add_album_map album_id {218} album_artist {8}
2023-08-25T14:16:27+01:00 [eric] (Ampache\Repository\Model\Art) -> duplicate... type:album old_id:217 new_type:album new_id:218
2023-08-25T14:16:27+01:00 [eric] (Ampache\Repository\Model\Song) -> media-diff {"album":"OLD: 217 --> 218","mode":"OLD:abr --> vbr"}
2023-08-25T14:16:27+01:00 [eric] (Ampache\Repository\Model\Catalog) -> /home/eric/Testing/orphan/ray_of_light/drowned_world__substitute_for_love.ogg : differences found, updating database
2023-08-25T14:16:27+01:00 [eric] (Ampache\Repository\Model\Catalog) -> migrate 1168 album: {217} to {218}
2023-08-25T14:16:27+01:00 [eric] (Ampache\Repository\Model\Art) -> duplicate... type:album old_id:217 new_type:album new_id:218
```

**Comment by Nizhile** [comment]
This might be related to https://github.com/JamesHeinrich/getID3/issues/214 . But using demo scripts from command line does not trigger any problem. Can this be related to php module?

**Comment by mezco2** [comment]
@Nizhile  - Hi again.
I'm not sure exactly what you are looking for, so I have attached another log file. 
(As an FYI, I have the music broken down into 28 directories: "A" through "Z", "The" and "Assorted". I have created 28 catalogs to match the directories. I did this because when I originally added all of my music, I noticed that the "orphaned" error was happening. Thinking that it might have been some corrupt files, I sorted the music as explained above, to no avail.)
The attached log file is representative of my logging in and updating the "L" catalog. I confirmed that the artists were correctly listed before I updated the catalog, and I confirmed that the Unknown (Orphaned) problem happened after I updated the catalog.
Thanks again for all of your help.
[ampache.20230827.log](https://github.com/ampache/ampache/files/12447031/ampache.20230827.log)

**Comment by Nizhile** [comment]
@mezco2  you lost information in your album because it is actually a new instance in your database. Your first Liz Phair album was created with id 446, then your verified operation create a new id 1178. The problem seems to be related to  `(Ampache\Repository\Model\Album) -> check album: created {1178}` line.
I fixed locally the bitrate mode issue for ogg, but the check_album continue to decide on a new album creation.
```
2023-08-27T00:56:30+00:00 [admin] (local.catalog) -> catalog 16 starting verify album on chunk 0
2023-08-27T00:56:30+00:00 [admin] (Ampache\Repository\Model\Catalog) -> Reading tags from /home/davemezc/music.davemez.com/app/music/L/Liz Phair/Whip-Smart [Single]/01-Whip-Smart [Remix].mp3
2023-08-27T00:56:30+00:00 [admin] (Ampache\Repository\Model\Album) -> check album: created {1178}
2023-08-27T00:56:30+00:00 [admin] (Ampache\Repository\Model\Catalog) -> update_map album: {1178}
2023-08-27T00:56:30+00:00 [admin] (Ampache\Repository\Model\Album) -> add_album_map album_id {1178} song_artist {287}
2023-08-27T00:56:30+00:00 [admin] (Ampache\Repository\Model\Artist) -> add_artist_map artist_id {152} album {1178}
2023-08-27T00:56:30+00:00 [admin] (Ampache\Repository\Model\Album) -> add_album_map album_id {1178} album_artist {152}
2023-08-27T00:56:30+00:00 [admin] (Ampache\Repository\Model\Art) -> duplicate... type:album old_id:446 new_type:album new_id:1178
2023-08-27T00:56:30+00:00 [admin] (Ampache\Repository\Model\Song) -> media-diff {"album":"OLD: 446 --> 1178"}
```

**Comment by Nizhile** [comment]
The Album::check() function was called with `$album_artist = artist.id('Unknown (Orphan)')`. This makes the search for an existing album unsuccessful. And thus a new album is created with the incorrect album artist.

**Comment by mezco2** [comment]
@Nizhile 
Thank you so much for your insight! You are way ahead of me in terms of understanding this.
Thanks to you, I think I understand the "why", but I don't know how to correct the issue. Is this something that I can do myself by editing one or more of the .php files or is this a developer issue?
As I mentioned above, I am all too happy to help in any way that I can to get this issue resolved. My site is not live yet, so I am able to do whatever is necessary.
Thank you again for all of your help!

**Comment by lachlan-00** [comment]
it could be the check function. there must an empty string in the tags being sent to the function.
![image](https://github.com/ampache/ampache/assets/1305249/fb4a4a38-8724-451a-a5c3-e77a8a39bafa)

That means that it's checking for an artist and it's either null or empty and then creating the additional artist.

this says to me it's the tags with an array that's got an empty string or something else extra in the tags making it send a check for nothing.

**Comment by codemaestro** [comment]
Hey all. My original issue that I posted about is slightly different. I don't think the files I've been importing and updating have any metadata tags encoded. When creating the new library I have the app build the album artist/title and track number/title by parsing the filename pattern. This activity goes badly when I add a new folder of files and ask for an **Update**.

**Comment by Nizhile** [comment]
The problem is located here https://github.com/ampache/ampache/blob/2607976d4443208d3bfc70159c1e06390bdc259d/src/Repository/Model/Catalog.php#L2662 
Using Artist::check() always return an artist, even if there is none found. See https://github.com/ampache/ampache/blob/2607976d4443208d3bfc70159c1e06390bdc259d/src/Repository/Model/Artist.php#L821 
Thus the next condition  `if (!$new_song->albumartist) {` to set up the new song album artist to `$song->albumartist` is never true. The Album check will always tries to find album with albumartist == 'Unknown'.
The problem will be visible on collection/media without an albumartist tag.
The block might be rewritten probably to something as follow
```
if ($is_upload_albumartist || !$albumartist) { // albumartist empty
  // no usable tag or $is_upload_albumartist
   $new_song->albumartist = $song->albumartist; // force it to song value
} else {
  // usable tag and not $is_upload_albumartist
  $new_song->albumartist = Artist::check($albumartist, $albumartist_mbid);
  if (!$new_song->albumartist) {
                $new_song->albumartist = $song->albumartist;
  }
}
```

**Comment by Nizhile** [comment]
> Hey all. My original issue that I posted about is slightly different. I don't think the files I've been importing and updating have any metadata tags encoded. When creating the new library I have the app build the album artist/title and track number/title by parsing the filename pattern. This activity goes badly when I add a new folder of files and ask for an **Update**.

That's the point : the add operation is able to build database information from directory pattern, tags aggregation, external database while the check (also performed during update) discards existing Ampache database content and use only tag albumartist (even if missing)

**Comment by mezco2** [comment]
Hi all,
I think we're all on the same page - I too am having the same issue as @codemaestro. Catalogs are initially created perfectly, but the artists become Unknown Orphaned when an Add or an Update is done to a catalog.
@Nizhile , I'd be happy to experiment with any changes that you would like me to implement. I just need detailed instructions on what you would like me to do.
Thank you all again!

**Comment by codemaestro** [comment]
@Nizhile - You are a master! This worked! Thank you!

**Comment by mezco2** [comment]
@Nizhile 
Alas, it didn't work for me. My luck.
I've attached my catalog.php file FYI. I'm at the point of giving up because I really don't understand the programming. I really do appreciate all of your help.
Thank you again.

**Comment by Nizhile** [comment]
@mezco2 
> I've attached my catalog.php file FYI

it disappears in the digital black hole? To confirm what triggers the problem, I limit the investigation to only three media files in a single directory. Then I create a new catalog with it, verify data in Ampache interface, launch catalog chek and verify data again.
In the log you already shared, you clearly have the same kind of issue as myself and codemaestro. But ours was triggerred by no albumartist tag. Maybe some other condition triggers the creation of the new album.

Can you uncomment the line shown in https://github.com/ampache/ampache/issues/3534#issuecomment-1690890501 ? Then check the L catalog and send logs again.

Also we're using Ampache 5.6.1 and 5.6.2.

**Comment by mezco2** [comment]
Hi @Nizhile 
I created a test directory with three media files and created the catalog. All was good. Then I updated the catalog, and all was still good. Then I added another directory with three more songs, updated the catalog and all was still good!
Given this, I am going to assume that something else is going on here. I am going to delete all of the catalogs and rebuild them. 
Do you know of any way to back up the album art before I delete the catalogs? I had to manually add a significant amount of art.
I have also done what you asked me to do regarding the L catalog and I have attached the log file. I've also attached the log file that was created after I updated the L catalog. (The L catalog update log is the larger of the two files, the one with the "a" suffix.)
Thank you again for your amazing help!
[ampache.20230901.log](https://github.com/ampache/ampache/files/12499794/ampache.20230901.log)
[ampache.20230901-a.log](https://github.com/ampache/ampache/files/12499798/ampache.20230901-a.log)

**Comment by Nizhile** [comment]
@mezco2 I got it. Your issue is the one that was fixed (your mp3s have empty ALBUMARTIST tag. With the test catalog, you can not reproduce the problem because well it’s fixed 😉. On your production catalog, once the album are tagged "unknown" I think they’ll stay as is. You’ll need to update album artist manually to the existing artist. Obviously, if you set up art, genre ... on unknown artist, you cannot just trash the catalogs and add them once again. I usually have art as a jpeg inside the song folder for old ripping. But I do not now how to export it from the database. Hope this can help.

**Comment by mezco2** [comment]
@Nizhile 
You're truly awesome!!
Thank you again for all of your help!

**Comment by codemaestro** [comment]
Question for @Nizhile: have these updates been merged into 6.0.2 or will
they be released in a future update?

Thanks for all your help!

**Comment by mezco2** [comment]
@Nizhile 
Here's an update just FYI...
I discovered how to export the artwork [here](https://github.com/ampache/ampache/wiki/cli-faq). Unfortunately, all the files that were extracted were zero bytes in size. Ugh.
In any case, using PHPmyAdmin, I deleted all instances of Unknown (Orphaned) in the album database and updated the catalogs. IT WORKED!! The only giveback is that the artwork will need to be updated for the affected albums. (Ampache probably now thinks that the artist has changed and needs new artwork.)
I just figured that I'd let you know that your fix was ultimately successful!
Thank you again for all of your help. There is no doubt in my mind whatsoever that I could not have done this without your help.
